ksqldb-engine/src/main/java/io/confluent/ksql/format/DefaultFormatInjector.java

```diff
@@ -0,0 +1,190 @@
+    }
+  }
+
+  private Optional<ConfiguredStatement<CreateSource>> injectForCreateStatement(
+      final ConfiguredStatement<CreateSource> original
+  ) {
+
+    final CreateSource statement = original.getStatement();
+    final CreateSourceProperties properties = statement.getProperties();
+
+    final Optional<FormatInfo> keyFormat = properties.getKeyFormat();
+    final Optional<FormatInfo> valueFormat = properties.getValueFormat();
+
+    if (keyFormat.isPresent() && valueFormat.isPresent()) {
+      return Optional.empty();
+    }
+
+    final KsqlConfig config = getConfig(original);
+    validateConfig(config, keyFormat, valueFormat);
+
+    final CreateSourceProperties injectedProps = properties.withFormats(
+        keyFormat.map(FormatInfo::getFormat)
+            .orElse(config.getString(KsqlConfig.KSQL_DEFAULT_KEY_FORMAT_CONFIG)),
+        valueFormat.map(FormatInfo::getFormat)
+            .orElse(config.getString(KsqlConfig.KSQL_DEFAULT_VALUE_FORMAT_CONFIG))
+    );
+    final CreateSource withFormats = statement.copyWith(
+        original.getStatement().getElements(),
+        injectedProps
+    );
+
+    final PreparedStatement<CreateSource> prepared = buildPreparedStatement(withFormats);
+    final ConfiguredStatement<CreateSource> configured = ConfiguredStatement
+        .of(prepared, original.getConfigOverrides(), original.getConfig());
+
+    return Optional.of(configured);
+  }
+
+  private static void validateConfig(
+      final KsqlConfig config,
+      final Optional<FormatInfo> keyFormat,
+      final Optional<FormatInfo> valueFormat
+  ) {
+    if (!keyFormat.isPresent()
+        && config.getString(KsqlConfig.KSQL_DEFAULT_KEY_FORMAT_CONFIG) == null) {
+      throw new KsqlException("Statement is missing the '"
+          + CommonCreateConfigs.KEY_FORMAT_PROPERTY + "' property from the WITH clause. "
+          + "Either provide one or set a default via the '"
+          + KsqlConfig.KSQL_DEFAULT_KEY_FORMAT_CONFIG + "' config.");
+    }
+    if (!valueFormat.isPresent()
+        && config.getString(KsqlConfig.KSQL_DEFAULT_VALUE_FORMAT_CONFIG) == null) {
+      throw new KsqlException("Statement is missing the '"
+          + CommonCreateConfigs.VALUE_FORMAT_PROPERTY + "' property from the WITH clause. "
+          + "Either provide one or set a default via the '"
+          + KsqlConfig.KSQL_DEFAULT_VALUE_FORMAT_CONFIG + "' config.");
+    }
+  }
+
+  private static KsqlConfig getConfig(final ConfiguredStatement<?> statement) {
+    return statement.getConfig().cloneWithPropertyOverwrite(statement.getConfigOverrides());
+  }
+
+  private static void validateLegacyFormatProperties(final ConfiguredStatement<?> statement) {
+    if (statement.getStatement() instanceof CreateSource) {
+      final CreateSource createStatement = (CreateSource) statement.getStatement();
+
+      if (createStatement.getProperties().getKeyFormat().isPresent()) {
+        throwKeyFormatDisabled(statement.getStatementText());
+      }
```

Comment:
As discussed above, let's only throw if non `KAFKA`.

Reply:
As discussed offline, this check to throw if `KEY_FORMAT` or `FORMAT` was provided without the feature flag enabled has been removed entirely. The engine is responsible for validating the provided format, in the context of the feature flag.